CLI ghc-options don't appear to work with haddock

Repro:

Preparation + using stack to get into an execution environment.
```
git clone https://github.com/ndmitchell/stack-3137
cd stack-3137/bob
stack exec --no-ghc-package-path -- bash
```

The rest is stack agnostic, should behave this way with the following versions:

```
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.2.2
$ runhaskell Setup.hs --version
runhaskell Setup.hs --version
Cabal library version 2.0.1.0
```

Then, normal configure + build with custom ghc options:

```
runhaskell Setup.hs configure --ghc-options -XMagicHash
runhaskell Setup.hs build
```

That bit works fine.  However, I get the following error for `runhaskell Setup.hs haddock -v`

```
/home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/bin/haddock --ghc-version
/home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/bin/ghc --print-libdir -XMagicHash
creating dist/build
creating dist/build/autogen
creating dist/build/autogen
Preprocessing library for bob-0.1.0.0..
Running Haddock on library for bob-0.1.0.0..
creating dist/doc/html/bob
dist/doc/html/bob/haddock-response10375-1.txt contents: <<<
--prologue=dist/doc/html/bob/haddock-prologue10375-0.txt
--dump-interface=dist/doc/html/bob/bob.haddock
--package-name=bob
--package-version=0.1.0.0
--hide=Paths_bob
--verbosity=1
--html
--read-interface=file:///home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/share/doc/ghc-8.2.2/html/libraries/base-4.10.1.0,/home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/share/doc/ghc-8.2.2/html/libraries/base-4.10.1.0/base.haddock
--read-interface=file:///home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/share/doc/ghc-8.2.2/html/libraries/ghc-prim-0.5.1.1,/home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/share/doc/ghc-8.2.2/html/libraries/ghc-prim-0.5.1.1/ghc-prim.haddock
--read-interface=file:///home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/share/doc/ghc-8.2.2/html/libraries/integer-gmp-1.0.1.0,/home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/share/doc/ghc-8.2.2/html/libraries/integer-gmp-1.0.1.0/integer-gmp.haddock
--odir=dist/doc/html/bob/
--title=bob-0.1.0.0:\ Short\ description\ of\ your\ package
--optghc=-fbuilding-cabal-package
--optghc=-O
--optghc=-outputdir
--optghc=dist/build
--optghc=-odir
--optghc=dist/build/tmp-10375
--optghc=-hidir
--optghc=dist/build/tmp-10375
--optghc=-stubdir
--optghc=dist/build/tmp-10375
--optghc=-i
--optghc=-idist/build
--optghc=-isrc
--optghc=-idist/build/autogen
--optghc=-idist/build/global-autogen
--optghc=-Idist/build/autogen
--optghc=-Idist/build/global-autogen
--optghc=-Idist/build
--optghc=-optP-D__HADDOCK_VERSION__=2181
--optghc=-optP-include
--optghc=-optPdist/build/autogen/cabal_macros.h
--optghc=-this-unit-id
--optghc=bob-0.1.0.0-5MPKpCkQZUc7bW0aMIjltN
--optghc=-hide-all-packages
--optghc=-Wmissing-home-modules
--optghc=-no-user-package-db
--optghc=-package-id
--optghc=base-4.10.1.0
--optghc=-XHaskell2010
-B/home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/lib/ghc-8.2.2
src/Lib.hs
dist/build/autogen/Paths_bob.hs
>>> dist/doc/html/bob/haddock-response10375-1.txt
/home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/bin/haddock '@dist/doc/html/bob/haddock-response10375-1.txt'
Haddock coverage:

src/Lib.hs:8:1: error:
    parse error (possibly incorrect indentation or mismatched brackets)
  |
8 | someFunc# = putStrLn "someFunc"
  | ^
```

Apparently, configured ghc-options are not passed into haddock.  Moreover, ghc-options passed in via `--ghc-options` CLI are also not passed in.

Using `--haddock-options -XMagicHash` doesn't work either, because it crashed when it tries to run `haddock --ghc-version -XMagicHash`.

It seems that it's been this way for some time:

```
$ stack exec --no-ghc-package-path --resolver lts-2.1 -- bash
$ runhaskell Setup.hs --version
Cabal library version 1.18.1.5
```

Has the same behavior for `haddock`, even if I do `runhaskell Setup.hs haddock --ghc-option="-XMagicHash"`, which should work according to `runhaskell Setup.hs haddock --help`.

--- a/Cabal/Distribution/Simple/Haddock.hs
+++ b/Cabal/Distribution/Simple/Haddock.hs
@@ -99,7 +99,7 @@ data HaddockArgs = HaddockArgs {
  -- ^ Page title, required.
  argPrologue :: Flag String,
  -- ^ Prologue text, required.
- argGhcOptions :: Flag (GhcOptions, Version),
+ argGhcOptions :: GhcOptions,
  -- ^ Additional flags to pass to GHC.
  argGhcLibDir :: Flag FilePath,
  -- ^ To find the correct GHC, required.
@@ -266,8 +266,12 @@ fromFlags env flags =
                       [ Hoogle | Flag True <- [haddockHoogle flags] ]
                  of [] -> [ Html ]
                     os -> os,
-      argOutputDir = maybe mempty Dir . flagToMaybe $ haddockDistPref flags
+      argOutputDir = maybe mempty Dir . flagToMaybe $ haddockDistPref flags,
+
+      argGhcOptions = mempty { ghcOptExtra = toNubListR ghcArgs }
     }
+    where
+      ghcArgs = fromMaybe [] . lookup "ghc" . haddockProgramArgs $ flags
 
 fromPackageDescription :: HaddockTarget -> PackageDescription -> HaddockArgs
 fromPackageDescription haddockTarget pkg_descr =
@@ -332,12 +336,9 @@ mkHaddockArgs verbosity tmp lbi clbi htmlTemplate haddockVersion inFiles bi = do
             then return sharedOpts
             else die' verbosity $ "Must have vanilla or shared libraries "
                        ++ "enabled in order to run haddock"
-    ghcVersion <- maybe (die' verbosity "Compiler has no GHC version")
-                        return
-                        (compilerCompatVersion GHC (compiler lbi))
 
     return ifaceArgs {
-      argGhcOptions  = toFlag (opts, ghcVersion),
+      argGhcOptions  = opts,
       argTargets     = inFiles
     }
 
@@ -555,7 +556,7 @@ renderPureArgs version comp platform args = concat
          id (getAny $ argIgnoreExports args))
       . fromFlag . argTitle $ args
 
-    , [ "--optghc=" ++ opt | (opts, _ghcVer) <- flagToList (argGhcOptions args)
+    , [ "--optghc=" ++ opt | let opts = argGhcOptions args
                            , opt <- renderGhcOptions comp platform opts ]
 
     , maybe [] (\l -> ["-B"++l]) $

